File: Leetcode_POTD_2025/10_OCTOBER/3494_Find_the_Minimum_Amount_of_Time_to_Brew_Potions.py
class Solution(object):
    def minTime(self, vertex, conduit):
        v_size = len(vertex)
        v_track = [0] * v_size
        
        for c_val in conduit:
            # 1. Base
            v_track[0] = v_track[0] + vertex[0] * c_val
            
            # 2. Forward Propagate
            for k in range(1, v_size):
                proc_time = vertex[k] * c_val
                v_track[k] = max(v_track[k], v_track[k - 1]) + proc_time
                
            # 3. Backward Adjust
            for k in range(v_size - 2, -1, -1):
                proc_time_next = vertex[k + 1] * c_val
                v_track[k] = v_track[k + 1] - proc_time_next
                
        return v_track[-1]

File: Leetcode_POTD_2025/10_OCTOBER/3494_Find_the_Minimum_Amount_of_Time_to_Brew_Potions.java
import java.util.Arrays;

class Solution {
    public long minTime(int[] vertex, int[] conduit) {
        int v_size = vertex.length;
        long[] v_track = new long[v_size];

        for (int c_val : conduit) {
            // 1. Base
            v_track[0] = v_track[0] + (long)vertex[0] * c_val;

            // 2. Forward Propagate
            for (int k = 1; k < v_size; k++) {
                long proc_time = (long)vertex[k] * c_val;
                v_track[k] = Math.max(v_track[k], v_track[k - 1]) + proc_time;
            }

            // 3. Backward Adjust
            for (int k = v_size - 2; k >= 0; k--) {
                long proc_time_next = (long)vertex[k + 1] * c_val;
                v_track[k] = v_track[k + 1] - proc_time_next;
            }
        }
        return v_track[v_size - 1];
    }
}
